### PR TITLE
Add ability to verify and restore metadata

### PR DESCRIPTION
**Pull request by legionus**
[x] Find broken imagestream tags that reference non-existent objects in the storage.
[x] Find broken images that reference non-existent objects in the storage.
[x] Restore imagestream based on the storage data.
[x] Restore imagestream tag (not tagname) based on the storage data.

These options require:
```
oc adm policy add-cluster-role-to-user cluster-admin system:serviceaccount:default:registry
```

### CODE REVIEW

File: pkg/cmd/dockerregistry/dockerregistry.go [hunk 1]
```diff
@@ -50,6 +50,10 @@ import (
 )
 
 var pruneMode = flag.String("prune", "", "prune blobs from the storage and exit (check, delete)")
+var listRepositories = flag.Bool("list-repositories", false, "shows list of repositories")
+var listBlobs = flag.Bool("list-blobs", false, "shows list of blob digests stored in the storage")
+var listManifests = flag.Bool("list-manifests", false, "shows list of manifest digests stored in the storage")
+var listRepositoryManifests = flag.String("list-repository-manifests", "", "shows the manifest digests in all the repository")
```
Comment (dmage) [inline pkg/cmd/dockerregistry/dockerregistry.go]: I'd suggest `list-manifests-from`

Comment (legionus) [reply pkg/cmd/dockerregistry/dockerregistry.go]: @dmage Fixed

File: pkg/imagestream/image.go [hunk 2]
```diff
@@ -11,15 +11,9 @@ import (
 	imageapiv1 "github.com/openshift/api/image/v1"
 
 	"github.com/openshift/image-registry/pkg/dockerregistry/server/client"
-	imageapi "github.com/openshift/image-registry/pkg/origin-common/image/apis/image"
 	util "github.com/openshift/image-registry/pkg/origin-common/util"
 )
 
-func IsImageManaged(image *imageapiv1.Image) bool {
```
[the remark this answers is not in the document]
Comment (legionus) [reply pkg/imagestream/image.go]: @dmage Fixed

File: pkg/imagestream/imagestream.go [hunk 3]
```diff
@@ -90,8 +92,8 @@ func (is *imageStream) Reference() string {
 	return fmt.Sprintf("%s/%s", is.namespace, is.name)
 }
 
-// createImageStream creates a new image stream and caches it.
-func (is *imageStream) createImageStream(ctx context.Context, userClient client.Interface) (*imageapiv1.ImageStream, error) {
+// CreateImageStream creates a new image stream and caches it.
```
[the remark this answers is not in the document]
Comment (legionus) [reply pkg/imagestream/imagestream.go]: @dmage Fixed

**Comment by legionus** [comment]
@bparees PTAL

File: pkg/dockerregistry/server/prune/restore.go [hunk 4]
```diff
@@ -0,0 +1,377 @@
...
+}
+
+var _ Restore = &StorageRestore{}
+
+func (r *StorageRestore) ImageStreamTag(imageStream imagestream.ImageStream, image *imageapiv1.Image, tagName string) error {
+	return imageStream.CreateImageStreamMapping(r.Ctx, r.Client, tagName, image)
+}
+
+func (r *StorageRestore) BrokenImage(image imageapiv1.Image, err error) error {
+	return nil
```
Comment (bparees) [inline pkg/dockerregistry/server/prune/restore.go]: shouldn't this at least print something, since it's not fixing it?

File: pkg/dockerregistry/server/prune/restore.go [hunk 5]
```diff
@@ -0,0 +1,377 @@
...
+}
+
+func (r *StorageRestore) BrokenImageStreamTag(is imageapiv1.ImageStream, tag imageapiv1.NamedTagEventList, pos int, err error) error {
+	return nil
```
Comment (bparees) [inline pkg/dockerregistry/server/prune/restore.go]: same here, shouldn't this at least print something, since it's not fixing it?

File: pkg/dockerregistry/server/prune/restore.go (hunk 4 above)
Comment (legionus) [reply pkg/dockerregistry/server/prune/restore.go]: I don't think so because for this, there is a check mode.

File: pkg/dockerregistry/server/prune/restore.go (hunk 5 above)
Comment (legionus) [reply pkg/dockerregistry/server/prune/restore.go]: I don't think so because for this, there is a check mode.

File: pkg/cmd/dockerregistry/dockerregistry.go [hunk 6]
```diff
@@ -43,13 +40,18 @@ import (
...
 	registryconfig "github.com/openshift/image-registry/pkg/dockerregistry/server/configuration"
 	"github.com/openshift/image-registry/pkg/dockerregistry/server/maxconnections"
-	"github.com/openshift/image-registry/pkg/dockerregistry/server/prune"
 	"github.com/openshift/image-registry/pkg/origin-common/clientcmd"
 	"github.com/openshift/image-registry/pkg/origin-common/crypto"
 	"github.com/openshift/image-registry/pkg/version"
 )
 
 var pruneMode = flag.String("prune", "", "prune blobs from the storage and exit (check, delete)")
+var restoreMode = flag.String("restore", "", "check data corruption or recover storage data if possible (valid values: check, check-database, check-storage, recover)")
```
[the remark this answers is not in the document]
Comment (legionus) [reply pkg/cmd/dockerregistry/dockerregistry.go]: @bparees Fixed.

File: pkg/cmd/dockerregistry/dockerregistry.go [hunk 7]
```diff
@@ -43,13 +40,18 @@ import (
...
+var restoreNamespace = flag.String("restore-namespace", "", "check and recover only specified namespace")
+var listRepositories = flag.Bool("list-repositories", false, "shows list of repositories")
+var listBlobs = flag.Bool("list-blobs", false, "shows list of blob digests stored in the storage")
+var listManifests = flag.Bool("list-manifests", false, "shows list of manifest digests stored in the storage")
+var listRepositoryManifests = flag.String("list-manifests-from", "", "shows the manifest digests in all the repository")
```
[the remark this answers is not in the document]
Comment (legionus) [reply pkg/cmd/dockerregistry/dockerregistry.go]: @bparees Fixed.

File: pkg/cmd/dockerregistry/dockerregistry.go [hunk 8]
```diff
@@ -58,76 +60,64 @@ func versionFields() map[interface{}]interface{} {
...
 	}
 
-	var pruner prune.Pruner
+	listOpts := getListOptions()
 
-	if dryRun {
-		pruner = &prune.DryRunPruner{}
-	} else {
-		pruner = &prune.RegistryPruner{StorageDriver: storageDriver}
+	if listOpts.Repositories || listOpts.Blobs || listOpts.Manifests {
```
[the remark this answers is not in the document]
Comment (legionus) [reply pkg/cmd/dockerregistry/dockerregistry.go]: We do not use custom modules to parse command line options and do not have a subcommand too. I added the `-experimental=true` option.

File: pkg/dockerregistry/server/prune/restore.go [hunk 9]
```diff
@@ -0,0 +1,377 @@
...
+	}
+
+	return nil
+}
+
+// Database checks metadata in the database
+func (r *Fsck) Database(namespace string) error {
+	listIS, err := r.Client.ImageStreams(namespace).List(metav1.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to list all image streams: %s", err)
```
[the remark this answers is not in the document]
Comment (legionus) [reply pkg/dockerregistry/server/prune/restore.go]: @bparees Fixed

File: pkg/dockerregistry/server/prune/restore.go [hunk 10]
```diff
@@ -0,0 +1,377 @@
...
+				DockerImageConfig:            string(config),
+				DockerImageLayers:            layers,
+			}
+
+			return r.Restore.ImageStreamTag(imageStream, image, "lost-found-"+dgst.Hex())
+		})
+		if _, ok := err.(driver.PathNotFoundError); ok {
+			return nil
+		} else if err != nil {
+			return fmt.Errorf("failed to prune manifest links in the repository %s: %v", repoName, err)
```
[the remark this answers is not in the document]
Comment (legionus) [reply pkg/dockerregistry/server/prune/restore.go]: @bparees Fixed

File: pkg/dockerregistry/server/prune/restore.go (hunk 4 above)
Comment (bparees) [reply pkg/dockerregistry/server/prune/restore.go]: So check mode will tell you "image X is broken" and normal execution mode will just print nothing, which would make someone think "ok it fixed the items check reported as broken" except it didn't?

How about something like "Fixing broken images not implemented, skipping broken image X"?

(same for imagestreamtags)

File: pkg/cmd/dockerregistry/dockerregistry.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (bparees) [reply pkg/cmd/dockerregistry/dockerregistry.go]: works for me, thanks

File: pkg/dockerregistry/server/prune/restore.go [hunk 11]
```diff
@@ -0,0 +1,380 @@
...
+}
+
+// Database checks metadata in the database
+func (r *Fsck) Database(namespace string) error {
+	listIS, err := r.Client.ImageStreams(namespace).List(metav1.ListOptions{})
+	if err != nil {
+		if namespace == metav1.NamespaceAll {
+			namespace = "all"
+		}
+		return fmt.Errorf("failed to list image streams in the %s namespace(s): %s", namespace, err)
```
Comment (bparees) [inline pkg/dockerregistry/server/prune/restore.go]: why (s) ?  seems like %s is only ever going to be a single namespace name.

**Comment by bparees** [comment]
@legionus one nit and lgtm code-wise.

one other thing i'd like to see from an R&D perspective is a readme/doc that covers:

1) what scenarios this attempts to address and what it does for each of them (e.g. talking about the lost+found imagestream).  Things like "when an image manifest contains a layer that cannot be found in storage...."

2) for cases where it currently does nothing beyond reporting the issue, some discussion of what we could do in the future, or if nothing can be done, why nothing can be done (and perhaps any recommendation for manual action to be taken in such cases).

I know that goes a little beyond a PoC but since it seems likely this may sit for a bit before we get to come back and GA it, it'd be good to make it as usable as possible as well as leaving us with a reminder of what we might want to do next in this space.

**Comment by bparees** [comment]
sorry, more than one nit.  still need to sort out out discussion of what "restore mode" should be printing when it takes no action on something that's broken.

File: pkg/dockerregistry/server/prune/restore.go (hunk 11 above)
Comment (legionus) [reply pkg/dockerregistry/server/prune/restore.go]: @bparees Because `failed to list image streams in the foo namespace` or `failed to list image streams in the all namespaces`. No ?

File: pkg/dockerregistry/server/prune/restore.go (hunk 4 above)
Comment (legionus) [reply pkg/dockerregistry/server/prune/restore.go]: @bparees OK. I will add some message.

File: pkg/dockerregistry/server/prune/restore.go (hunk 11 above)
Comment (bparees) [reply pkg/dockerregistry/server/prune/restore.go]: Hm, well it's a bit awkward("in the all namespaces"), how about rewording to:
`failed to list image streams in namespace(s): %s, error: %s`

**Comment by legionus** [comment]
@bparees please review again

File: pkg/dockerregistry/server/prune/restore.go (hunk 11 above)
Comment (bparees) [reply pkg/dockerregistry/server/prune/restore.go]: @legionus you may have missed this comment

File: pkg/dockerregistry/server/prune/doc.go [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+// 1. Show broken imagestream tags that refer to non-existent images on the storage.
+// In this case, you can either delete such a tag, or re-push the image pointed to by this tag.
+//
+// 2. Show broken images that refer to non-existent blobs on the storage.  In this case,
+// you can re-push any imagestream tag that point to this image. Since the images are global,
+// it's enough to re-push only one image to fix all the imagestream tags that refer to it.
+//
+// 3. Show and restore imagestreams based on the storage data. All tags inside will
+// have the form: lost-found-<IMAGE-DIGEST> because the name of the tag is not stored
+// on the storage and can't be restored. You can make a new tag with a different name.
```
Comment (bparees) [inline pkg/dockerregistry/server/prune/doc.go]: can you elaborate on this a little?  I think what this is doing is taking repositories that exist in storage (but aren't referenced by an imagestream in etcd) and creating imagestreams in etcd (plus lost+found tags), but it's not totally clear.

Will this create the entire imagestream if it was deleted, or only add lost+found tags to an existing imagestream?

File: pkg/dockerregistry/server/prune/doc.go [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+//
+// 4. Show and restore imagestream tag (not a tag name itself) based on the storage data.
+// After that you can change the tag name to the correct one.
```
Comment (bparees) [inline pkg/dockerregistry/server/prune/doc.go]: how is this different from (3)?  i'm not sure what "imagestream tag (not a tag name itself)" means?

**Comment by bparees** [comment]
couple questions on the doc but mostly looks good

File: pkg/dockerregistry/server/prune/restore.go (hunk 11 above)
Comment (legionus) [reply pkg/dockerregistry/server/prune/restore.go]: @bparees oops. sorry.

File: pkg/dockerregistry/server/prune/doc.go (hunk 12 above)
Comment (legionus) [reply pkg/dockerregistry/server/prune/doc.go]: @bparees fsck will restore the entire imagestream with all imagestream tags if it was deleted.
I corrected the `doc.go`. PTAL.

File: pkg/dockerregistry/server/prune/doc.go (hunk 13 above)
Comment (legionus) [reply pkg/dockerregistry/server/prune/doc.go]: @bparees Fixed

File: pkg/dockerregistry/server/prune/doc.go [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+// This mode covers the following cases:
+//
+// 1. Show broken imagestream tags that refer to non-existent images on the storage.
+// In this case, you can either delete such a tag, or re-push the image pointed to by this tag.
+//
+// 2. Show broken images that refer to non-existent blobs on the storage.  In this case,
+// you can re-push any imagestream tag that point to this image. Since the images are global,
+// it's enough to re-push only one image to fix all the imagestream tags that refer to it.
+//
+// 3. Show and restore either entire imagestreams or individual imagestram tags inside it
```
Comment (bparees) [inline pkg/dockerregistry/server/prune/doc.go]: s/imagestram/imagestream/

Comment (legionus) [reply pkg/dockerregistry/server/prune/doc.go]: @bparees Fixed

**Comment by legionus** [comment]
> one typo and lgtm but again due to the prune changes i think we should hold this to post-3.9 so QE can do a full regression. 

@bparees as you wish. But if you look closer, you'll see that I just unified the code.
/hold

**Comment by bparees** [comment]
> @bparees as you wish. But if you look closer, you'll see that I just unified the code.

ok i'll take another pass this afternoon

**Comment by bparees** [comment]
@legionus ok, you're right the prune changes seem pretty straightforward.  squash+merge
